### PR TITLE
Disregard leading zeros when comparing package version numbers with Pip

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/1401)

### CODE REVIEW

File: analyzer/src/main/kotlin/managers/PIP.kt [hunk 1]
```diff
@@ -403,31 +403,31 @@ class PIP(name: String, analyzerConfig: AnalyzerConfiguration, repoConfig: Repos
...
+        val url = binaryArtifact["url"]?.textValue() ?: pkg.binaryArtifact.url
+        val hash = binaryArtifact["md5_digest"]?.textValue() ?: pkg.binaryArtifact.hash
 
-        return RemoteArtifact(url = url, hash = hash, hashAlgorithm = HashAlgorithm.fromHash(hash))
+        return RemoteArtifact(url, hash, HashAlgorithm.fromHash(hash))
     }
 
-    private fun getSourceArtifact(pkgReleases: ArrayNode): RemoteArtifact {
-        val pkgSources = pkgReleases.asSequence().filter {
+    private fun getSourceArtifact(artifacts: ArrayNode): RemoteArtifact {
```
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/PIP.kt]: Considering only the those two function the change is fine.
The rename of the second parameter however makes the code style a bit inconsistent with the
local code style, e.g. the caller of `getSourceArtifact` still calls the node `pkgReleases`.
What I like about that is that it somehow indicates what 'JSON' node it expects.

How about (1) using 'releaseNode' as param name and also renaming the variable upwards the call tree?

Another alternative could be to pass the 'parent' node (the package one) to these function - it has the drawback that the function gets more than it needs and the advantage that always the package node could be passed and thus achieve more clarity about which JSON node were dealing with.

File: analyzer/src/main/kotlin/managers/PIP.kt [hunk 2]
```diff
@@ -153,6 +154,18 @@ class PIP(name: String, analyzerConfig: AnalyzerConfiguration, repoConfig: Repos
             "pypi.org",
             "pypi.python.org" // Legacy
         ).flatMap { listOf("--trusted-host", it) }.toTypedArray()
+
+        /**
+         * Return a version string with leading zeros of components stripped.
+         */
+        fun stripLeadingZerosFromVersion(version: String) =
```
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/PIP.kt]: Does this `version` parameter only contain digits and dots or are there valid cases where it contains additional characters. E.g. for artifacts "${name}-${version}-{classifier}" with non empty classifiers?

An option could be to: try/catch for each subsequence (between two dots) separately.

File: analyzer/src/main/kotlin/managers/PIP.kt (hunk 1 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/PIP.kt]: > The rename of the second parameter however makes the code style a bit inconsistent with the local code style, e.g. the caller of `getSourceArtifact` still calls the node `pkgReleases`.

Correct, that actually was part of the intention. I don't particularly like the variable-naming-style the caller uses, with prefixing all variables with `pkg`. If everything is `pkg` anyway, you may as well drop the prefix.

> What I like about that is that it somehow indicates what 'JSON' node it expects. [...] How about (1) using 'releaseNode' as param name [...]

Yes, that makes sense to me, will do.

> [...] and also renaming the variable upwards the call tree?

I'd prefer not to do that, at least not as part of this change, whose scope is limited to these sibling-functions.

Comment (fviernau) [reply analyzer/src/main/kotlin/managers/PIP.kt]: I don't like the `pkg` prefix either, my point was mainly consistent style within one file - however your proposed compromise makes sense as well.

File: analyzer/src/main/kotlin/managers/PIP.kt (hunk 2 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/PIP.kt]: The version string surely may contain other characters than digits and dots, see https://www.python.org/dev/peps/pep-0440/. But that's nothing the current implementation should struggle with...

> An option could be to: try/catch for each subsequence (between two dots) separately.

That's what the code already does. I guess you're confused by `split(".").joinToString(".")` which at first looks like a no-op, but the code following that line is an implicit mapping on the individual parts between two dots.

Comment (fviernau) [reply analyzer/src/main/kotlin/managers/PIP.kt]: I see, silly me - sorry then

**Comment by sschuberth** [comment]
AppVeyor seems to have some issues as it does not complete. Still merging as Travis completed successfully.

**Pull request merged**